Haskell/Core/Simplicity/Term/Core.hs

```diff
@@ -105,6 +107,22 @@ iioh = drop ioh
 iiih :: (Core term, TyC a, TyC b, TyC c, TyC x) => term (a, (b, (c, x))) x
 iiih = drop iih
 
+-- | Term for swapping positions in products (Commutativity of Multiplication): A x B |- B x A
+swapP :: (Core term, TyC a, TyC b) => term (a, b) (b, a)
+swapP = pair (drop iden) (take iden)
+
+-- | Term for swapping positions in sums (Commutativity of Addition): A + B |- B + A
+swapS :: (Core term, TyC a, TyC b) => term (Either a b) (Either b a)
+swapS = iden &&& unit >>> (match (take (injr iden)) (take (injl iden)))
+
+-- | Term for distributing products over sums (product of sums "CNF" to sum of products "DNF")
+cnf2dnf :: (Core term, TyC a, TyC b, TyC c) => term (a, Either b c) (Either (a, b) (a, c))
+cnf2dnf = swapP >>> (match (injl swapP) (injr swapP))
+
+-- | Term for factoring common products out of sums (sum of products "DNF" to product of sums "CNF")
+dnf2cnf :: (Core term, TyC a, TyC b, TyC c) => term (Either (a, b) (a, c)) (a, Either b c)
+dnf2cnf = iden &&& unit >>> match (ooh &&& injl oih) (ooh &&& injr oih)
```

Comment:
`iden &&& unit >>> match (take (oh &&& injl ih)) (take (oh &&& injr ih))` is slightly better.

Reply:
for conciseness, performance, canonical representation?

Reply:
Is it both more concise, which is probably the most important reason, and there is a fractional performance improvement since you only execute take once instead of twice.